#### Loading SQL extension

In [1]:
%load_ext sql

### Loading database file

In [2]:
%sql mysql+pymysql://root:*****@localhost:3306/md_water_services

'Connected: root@md_water_services'

#### Viewing Data Dictionary

Before we start, scan through the data dictionary, and perhaps query a couple of tables to get a feel for the database again.

In [7]:
%%sql
SELECT
    *
FROM 
    md_water_services.data_dictionary
LIMIT 100;        

 * mysql+pymysql://root:***@localhost:3306/md_water_services
   mysql+pymysql://root:***@localhost:3306/united_nations
49 rows affected.


[('employee', 'assigned_employee_id', 'Unique ID assigned to each employee', 'INT', 'visits'),
 ('employee', 'employee_name', 'Name of the employee', 'VARCHAR(255)', ''),
 ('employee', 'phone_number', 'Contact number of the employee', 'VARCHAR(15)', ''),
 ('employee', 'email', 'Email address of the employee', 'VARCHAR(255)', ''),
 ('employee', 'address', 'Residential address of the employee', 'VARCHAR(255)', ''),
 ('employee', 'town_name', 'Name of the town where the employee resides', 'VARCHAR(255)', ''),
 ('employee', 'province_name', 'Name of the province where the employee resides', 'VARCHAR(255)', ''),
 ('employee', 'position', 'Position or job title of the employee', 'VARCHAR(255)', ''),
 ('visits', 'record_id', 'Unique ID assigned to each visit', 'int', 'water_quality, water_source'),
 ('visits', 'location_id', 'ID of the location visited', 'varchar(255)', 'location'),
 ('visits', 'source_id', 'ID of the water source visited', 'varchar(510)', 'well_pollution'),
 ('visits', 'time_of_record', 'Date and time of the visit', 'datetime', ''),
 ('visits', 'visit_count', 'Number of visits made to this location', 'int', ''),
 ('visits', 'time_in_queue', 'Time spent by people waiting for water in a queue at the location', 'int', ''),
 ('visits', 'assigned_employee_id', 'ID of the employee who visited the location', 'int', 'employee'),
 ('water_quality', 'record_id', 'Unique ID assigned to each record', 'int', ''),
 ('water_quality', 'subjective_quality_score', 'Score representing the subjective quality of the water source', 'int', ''),
 ('water_quality', 'visit_count', 'Number of visits made for data collection', 'int', ''),
 ('water_source', 'source_id', 'Unique ID assigned to each water source', 'VARCHAR(510)', 'visits'),
 ('water_source', 'type_of_water_source', 'Type or category of the water source. Can be : tap_in_home, tap_in_home_broken, well, shared_tap, river', 'VARCHAR(255)', ''),
 ('water_source', 'number_of_people_served', 'Number of people served by this water source', 'INT', ''),
 ('well_pollution', 'source_id', 'ID of the water source being tested for pollution', 'varchar(258)', 'visits'),
 ('well_pollution', 'date', 'Date of the pollution test', 'datetime', ''),
 ('well_pollution', 'description', 'Description of the pollution test', 'varchar(255)', ''),
 ('well_pollution', 'pollutant_ppm', 'Result of the pollution test in parts per million', 'float', ''),
 ('well_pollution', 'biological', 'Biological contamination level', 'float', ''),
 ('well_pollution', 'results', "Result of the pollution test. Can be 'dirty', 'biologically contaminated', or 'clean'", 'varchar(255)', ''),
 ('location', 'location_id', 'Unique ID assigned to each location', 'varchar(255)', 'visits'),
 ('location', 'address', 'Address of the location', 'varchar(255)', ''),
 ('location', 'province_name', 'Name of the province where the location is situated', 'varchar(255)', ''),
 ('location', 'town_name', 'Name of the town where the location is situated', 'varchar(255)', ''),
 ('location', 'location_type', 'Type or category of the location', 'varchar(255)', ''),
 ('global_water_access', 'name', 'The country or area name', 'varchar(255)', ''),
 ('global_water_access', 'region', 'Geographical region', 'varchar(255)', ''),
 ('global_water_access', 'year', 'Year of the data record', 'int', ''),
 ('global_water_access', 'pop_n', 'The national population size estimate in thousands', 'float', ''),
 ('global_water_access', 'pop_u', 'The urban population share estimate in percentage points (%)', 'float', ''),
 ('global_water_access', 'wat_bas_n', 'The estimated national share of people with at least basic service (%)', 'float', ''),
 ('global_water_access', 'wat_lim_n', 'The estimated national share of people with limited service (%)', 'float', ''),
 ('global_water_access', 'wat_unimp_n', 'The estimated national share of people with unimproved service (%)', 'float', ''),
 ('global_water_access', 'wat_sur_n', 'The estimated national share of people with surfa

#### Cleaning our data
Ok, bringing up the employee table. It has info on all of our workers, but note that the email addresses have not been added. We will have to send
them reports and figures, so let's update it. Luckily the emails for our department are easy: first_name.last_name@ndogowater.gov.

In [11]:
%%sql
SELECT
    employee_name,
    CONCAT(LOWER(REPLACE(employee_name, " ", ".")),'@ndogowater.gov') AS new_email
FROM
    md_water_services.employee
LIMIT 10;    
        

 * mysql+pymysql://root:***@localhost:3306/md_water_services
   mysql+pymysql://root:***@localhost:3306/united_nations
10 rows affected.


[('Amara Jengo', 'amara.jengo@ndogowater.gov'),
 ('Bello Azibo', 'bello.azibo@ndogowater.gov'),
 ('Bakari Iniko', 'bakari.iniko@ndogowater.gov'),
 ('Malachi Mavuso', 'malachi.mavuso@ndogowater.gov'),
 ('Cheche Buhle', 'cheche.buhle@ndogowater.gov'),
 ('Zuriel Matembo', 'zuriel.matembo@ndogowater.gov'),
 ('Deka Osumare', 'deka.osumare@ndogowater.gov'),
 ('Lalitha Kaburi', 'lalitha.kaburi@ndogowater.gov'),
 ('Enitan Zuri', 'enitan.zuri@ndogowater.gov'),
 ('Farai Nia', 'farai.nia@ndogowater.gov')]

We have to update the database again with these email addresses, so before we do, let's use a SELECT query to get the format right, then use
UPDATE and SET to make the changes.


In [14]:
%%sql
UPDATE employee
SET email = CONCAT(LOWER(REPLACE(employee_name, ' ', '.')),'@ndogowater.gov')

 * mysql+pymysql://root:***@localhost:3306/md_water_services
   mysql+pymysql://root:***@localhost:3306/united_nations
56 rows affected.


ResourceClosedError: This result object does not return rows. It has been closed automatically.

In [5]:
%%sql
SELECT
    *
FROM
    md_water_services.employee  
LIMIT 10;      

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[(0, 'Amara Jengo', '+99637993287', 'amara.jengo@ndogowater.gov', '36 Pwani Mchangani Road', 'Sokoto', 'Ilanga', 'Field Surveyor'),
 (1, 'Bello Azibo', '+99643864786', 'bello.azibo@ndogowater.gov', '129 Ziwa La Kioo Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (2, 'Bakari Iniko', '+99222599041', 'bakari.iniko@ndogowater.gov', '18 Mlima Tazama Avenue', 'Hawassa', 'Rural', 'Field Surveyor'),
 (3, 'Malachi Mavuso', '+99945849900', 'malachi.mavuso@ndogowater.gov', '100 Mogadishu Road', 'Akatsi', 'Lusaka', 'Field Surveyor'),
 (4, 'Cheche Buhle', '+99381679640', 'cheche.buhle@ndogowater.gov', '1 Savanna Street', 'Akatsi', 'Rural', 'Field Surveyor'),
 (5, 'Zuriel Matembo', '+99034075111', 'zuriel.matembo@ndogowater.gov', '26 Bahari Ya Faraja Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (6, 'Deka Osumare', '+99379364631', 'deka.osumare@ndogowater.gov', '104 Kenyatta Street', 'Akatsi', 'Rural', 'Field Surveyor'),
 (7, 'Lalitha Kaburi', '+99681623240', 'lalitha.kaburi@ndogowater.gov', '145 Sungura Amanpour Road', 'Kilimani', 'Rural', 'Field Surveyor'),
 (8, 'Enitan Zuri', '+99248509202', 'enitan.zuri@ndogowater.gov', '117 Kampala Road', 'Hawassa', 'Zanzibar', 'Field Surveyor'),
 (10, 'Farai Nia', '+99570082739', 'farai.nia@ndogowater.gov', '33 Angélique Kidjo Avenue', 'Amanzi', 'Dahabu', 'Field Surveyor')]

I picked up another bit we have to clean up. Often when databases are created and updated, or information is collected from different sources,
errors creep in. For example, if you look at the phone numbers in the `phone_number` column, the values are stored as strings

In [20]:
%%sql
SELECT
    LENGTH(phone_number),
    phone_number
FROM
    employee
GROUP BY
    phone_number    
LIMIT 10;        

 * mysql+pymysql://root:***@localhost:3306/md_water_services
   mysql+pymysql://root:***@localhost:3306/united_nations
10 rows affected.


[(13, '+99637993287 '),
 (13, '+99643864786 '),
 (13, '+99222599041 '),
 (13, '+99945849900 '),
 (13, '+99381679640 '),
 (13, '+99034075111 '),
 (13, '+99379364631 '),
 (13, '+99681623240 '),
 (13, '+99248509202 '),
 (13, '+99570082739 ')]

The phone numbers should be 12 characters long, consisting of the plus sign, area code (99), and the phone number digits. However, when we use
the `LENGTH(column)` function, it returns **13** characters, indicating there's an extra character

In [21]:
%%sql
UPDATE employee
SET phone_number = TRIM(phone_number)

 * mysql+pymysql://root:***@localhost:3306/md_water_services
   mysql+pymysql://root:***@localhost:3306/united_nations
56 rows affected.


ResourceClosedError: This result object does not return rows. It has been closed automatically.

In [28]:
%%sql
SELECT
    LENGTH(phone_number) AS 'Phone number length',
    phone_number
FROM
    employee
GROUP BY
    phone_number    
LIMIT 10; 

 * mysql+pymysql://root:***@localhost:3306/md_water_services
   mysql+pymysql://root:***@localhost:3306/united_nations
10 rows affected.


[(12, '+99637993287'),
 (12, '+99643864786'),
 (12, '+99222599041'),
 (12, '+99945849900'),
 (12, '+99381679640'),
 (12, '+99034075111'),
 (12, '+99379364631'),
 (12, '+99681623240'),
 (12, '+99248509202'),
 (12, '+99570082739')]

In [27]:
%%sql
SELECT
    town_name,
    COUNT(DISTINCT(employee_name)) AS "Number of employees"
FROM
    employee
GROUP BY
    town_name


 * mysql+pymysql://root:***@localhost:3306/md_water_services
   mysql+pymysql://root:***@localhost:3306/united_nations
9 rows affected.


[('Dahabu', 6),
 ('Harare', 5),
 ('Ilanga', 3),
 ('Kintampo', 1),
 ('Lusaka', 4),
 ('Rural', 29),
 ('Serowe', 3),
 ('Yaounde', 1),
 ('Zanzibar', 4)]

In [29]:
%%sql
SELECT
    employee_name, e.email, e.phone_number, COUNT(*) AS visit_count
FROM
    md_water_services.employee e
JOIN
    md_water_services.visits v ON e.assigned_employee_id = v.assigned_employee_id
GROUP BY 
    e.assigned_employee_id
ORDER BY
    visit_count DESC
LIMIT 3;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
3 rows affected.


[('Bello Azibo', 'bello.azibo@ndogowater.gov', '+99643864786', 3708),
 ('Pili Zola', 'pili.zola@ndogowater.gov', '+99822478933', 3676),
 ('Rudo Imani', 'rudo.imani@ndogowater.gov', '+99046972648', 3539)]

In [48]:
%%sql
SELECT
    town_name, COUNT(town_name) AS num_of_records
FROM
    Location
GROUP BY 
    town_name
ORDER BY
    num_of_records DESC    
LIMIT 100;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
25 rows affected.


[('Rural', 23740),
 ('Harare', 1650),
 ('Amina', 1090),
 ('Lusaka', 1070),
 ('Mrembo', 990),
 ('Asmara', 930),
 ('Dahabu', 930),
 ('Kintampo', 780),
 ('Ilanga', 780),
 ('Isiqalo', 770),
 ('Zuri', 750),
 ('Amara', 710),
 ('Majengo', 620),
 ('Pwani', 520),
 ('Kofi', 510),
 ('Serowe', 480),
 ('Bahari', 470),
 ('Marang', 460),
 ('Abidjan', 400),
 ('Bello', 400),
 ('Cheche', 370),
 ('Deka', 350),
 ('Djenne', 310),
 ('Zanzibar', 300),
 ('Yaounde', 270)]

In [54]:
%%sql
SELECT
    province_name, COUNT(*) AS num_of_records
FROM
    Location
GROUP BY
    province_name
ORDER BY
    num_of_records DESC;    


 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('Kilimani', 9510),
 ('Akatsi', 8940),
 ('Sokoto', 8220),
 ('Amanzi', 6950),
 ('Hawassa', 6030)]

From this table, it's pretty clear that most of the water sources in the survey are situated in small rural communities, scattered across Maji Ndogo.
If we count the records for each province, most of them have a similar number of sources, so every province is well-represented in the survey

In [52]:
%%sql
SELECT
    province_name,
    town_name,
    COUNT(town_name) AS records_per_town
FROM
    Location
GROUP BY
    province_name, town_name
ORDER BY
    records_per_town DESC            


 * mysql+pymysql://root:***@localhost:3306/md_water_services
31 rows affected.


[('Akatsi', 'Rural', 6290),
 ('Kilimani', 'Rural', 5440),
 ('Sokoto', 'Rural', 5010),
 ('Hawassa', 'Rural', 3900),
 ('Amanzi', 'Rural', 3100),
 ('Akatsi', 'Lusaka', 1070),
 ('Kilimani', 'Mrembo', 990),
 ('Amanzi', 'Asmara', 930),
 ('Amanzi', 'Dahabu', 930),
 ('Kilimani', 'Harare', 850),
 ('Akatsi', 'Harare', 800),
 ('Akatsi', 'Kintampo', 780),
 ('Sokoto', 'Ilanga', 780),
 ('Kilimani', 'Isiqalo', 770),
 ('Kilimani', 'Zuri', 750),
 ('Kilimani', 'Amara', 710),
 ('Amanzi', 'Amina', 670),
 ('Sokoto', 'Majengo', 620),
 ('Amanzi', 'Pwani', 520),
 ('Sokoto', 'Kofi', 510),
 ('Hawassa', 'Serowe', 480),
 ('Sokoto', 'Bahari', 470),
 ('Sokoto', 'Marang', 460),
 ('Hawassa', 'Amina', 420),
 ('Amanzi', 'Abidjan', 400),
 ('Amanzi', 'Bello', 400),
 ('Sokoto', 'Cheche', 370),
 ('Hawassa', 'Deka', 350),
 ('Hawassa', 'Djenne', 310),
 ('Hawassa', 'Zanzibar', 300),
 ('Hawassa', 'Yaounde', 270)]

These results show us that our field surveyors did an excellent job of documenting the status of our country's water crisis. Every province and town
has many documented sources.
This makes me confident that the data we have is reliable enough to base our decisions on. This is an insight we can use to communicate data
integrity, so let's make a note of that

In [53]:
%%sql
SELECT
    location_type,
    COUNT(*) AS num_sources
FROM
    Location
GROUP BY 
    location_type
ORDER BY 
    num_sources DESC



 * mysql+pymysql://root:***@localhost:3306/md_water_services
2 rows affected.


[('Rural', 23740), ('Urban', 15910)]

We can see that there are more rural sources than urban, but it's really hard to understand those numbers. Percentages are more relatable.
If we use SQL as a very overpowered calculator:

In [58]:
%%sql
SELECT
    (23740/(23740 + 15910))*100 AS 'rural percentage'


 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(Decimal('59.8739'),)]

We can see that **60%** of all water sources in the data set are in rural communities.


So again, what are some of the insights we gained from the location table?
1. Our entire country was properly canvassed, and our dataset represents the situation on the ground.
2. 60% of our water sources are in rural communities across Maji Ndogo. We need to keep this in mind when we make decisions.

#### Diving into the sources
Ok, water_source is a big table, with lots of stories to tell, so strap in

The way I look at this table; we have access to different water source types and the number of people using each source.
These are the questions that I am curious about.
1. How many people did we survey in total?
2. How many wells, taps and rivers are there?
3. How many people share particular types of water sources on average?
4. How many people are getting water from each type of source?

In [62]:
%%sql
-- How many people did we survey in total?
SELECT
    SUM(number_of_people_served) AS Total_served_people
FROM
    water_source;        

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(Decimal('27628140'),)]

In [63]:
%%sql
-- How many wells, taps and rivers are there?
SELECT
    type_of_water_source, COUNT(*) AS number_of_sources
FROM
    water_source
GROUP BY  
    type_of_water_source
ORDER BY
    number_of_sources DESC;    



 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('well', 17383),
 ('tap_in_home', 7265),
 ('tap_in_home_broken', 5856),
 ('shared_tap', 5767),
 ('river', 3379)]

This isn't just an
informative result, we will need these numbers to understand how much all of these repairs will cost. If we know how many taps we need to install,
and we know how much it will cost to install them, we can calculate how much it will cost to solve the water crisis

In [65]:
%%sql
-- How many people share particular types of water sources on average?
SELECT 
    type_of_water_source,
    ROUND(AVG(number_of_people_served)) AS avg_num_of_people
FROM
    water_source
GROUP BY
    type_of_water_source
ORDER BY
    avg_num_of_people DESC;    



 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('shared_tap', Decimal('2071')),
 ('river', Decimal('699')),
 ('tap_in_home_broken', Decimal('649')),
 ('tap_in_home', Decimal('644')),
 ('well', Decimal('279'))]

These results are telling us that 644 people share a tap_in_home on average. Does that make sense?

No it doesn’t, The surveyors combined the data of many
households together and added this as a single tap record, but each household actually has its own tap. In addition to this, there is an average of
6 people living in a home. So 6 people actually share 1 tap (not 644)

This means that 1 tap_in_home actually represents 644 ÷ 6 = ± 100 taps.

Calculating the average number of people served by a single instance of each water source type helps us understand the typical capacity or load
on a single water source. This can help us decide which sources should be repaired or upgraded, based on the average impact of each upgrade.
For example, wells don't seem to be a problem, as fewer people are sharing them.

On the other hand, 2000 share a single public tap on average! We saw some of the queue times last time, and now we can see why. So looking at
these results, we probably should focus on improving shared taps first.

In [67]:
%%sql
-- How many people are getting water from each type of source?
SELECT
    type_of_water_source,
    SUM(number_of_people_served) AS total_people
from
    water_source
GROUP BY
    type_of_water_source  
ORDER BY
    total_people DESC          

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('shared_tap', Decimal('11945272')),
 ('well', Decimal('4841724')),
 ('tap_in_home', Decimal('4678880')),
 ('tap_in_home_broken', Decimal('3799720')),
 ('river', Decimal('2362544'))]

It's a little hard to comprehend these numbers, but you can see that one of these is dominating. To make it a bit simpler to interpret, let's use
percentages. First, we need the total number of citizens then use the result of that and divide each of the `SUM(number_of_people_served)` by
that number, times 100, to get percentages.

In [9]:
%%sql
SELECT
    type_of_water_source,
    ROUND((number_of_people_served / total_people_served * 100),2) AS percentage_served
FROM
    (SELECT
        type_of_water_source,
        SUM(number_of_people_served) AS number_of_people_served,
        (SELECT SUM(number_of_people_served) FROM water_source) AS total_people_served
     FROM
        water_source
     GROUP BY
        type_of_water_source) AS subquery
ORDER BY
    percentage_served DESC;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('shared_tap', Decimal('43.24')),
 ('well', Decimal('17.52')),
 ('tap_in_home', Decimal('16.94')),
 ('tap_in_home_broken', Decimal('13.75')),
 ('river', Decimal('8.55'))]

43% of our people are using `shared taps` in their communities, and on average, we saw earlier, that 2000 people share one shared_tap

By adding tap_in_home and tap_in_home_broken together, we see that 31% of people have water infrastructure installed in their homes, but 45%
(14/31) of these taps are not working! This isn't the tap itself that is broken, but rather the infrastructure like treatment plants, reservoirs, pipes, and
pumps that serve these homes that are broken.
18% of people are using wells. But only 4916 out of 17383 are clean = 28% (from last week).


**Start of a solution**
At some point, we will have to fix or improve all of the infrastructure, so we should start thinking about how we can make a data-driven decision
how to do it. I think a simple approach is to fix the things that affect most people first. So let's write a query that ranks each type of source based
on how many people in total use it. RANK() should tell you we are going to need a window function to do this, so let's think through the problem

We will need the following columns:
- Type of sources -- Easy
- Total people served grouped by the types -- We did that earlier, so that's easy too.
- A rank based on the total people served, grouped by the types -- A little harder

In [33]:
%%sql

SELECT
    type_of_water_source,
    SUM(number_of_people_served) AS total_people,
    RANK() OVER (ORDER BY SUM(number_of_people_served) DESC ) AS rank_by_population
    
FROM
    water_source
GROUP BY 
    type_of_water_source 
ORDER BY
    total_people DESC;           
          

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


[('shared_tap', Decimal('11945272'), 1),
 ('well', Decimal('4841724'), 2),
 ('tap_in_home', Decimal('4678880'), 3),
 ('tap_in_home_broken', Decimal('3799720'), 4),
 ('river', Decimal('2362544'), 5)]

Ok, so we should fix shared taps first, then wells, and so on. But the next question is, which shared taps or wells should be fixed first? We can use
the same logic; the most used sources should really be fixed first

In [56]:
%%sql
SELECT
    source_id,
    type_of_water_source,
    SUM(number_of_people_served) AS total_people,
    RANK() OVER (ORDER BY SUM(number_of_people_served) DESC) AS rank_priority
FROM
    water_source
GROUP BY
    source_id, type_of_water_source
ORDER BY
    rank_priority
LIMIT 100;
      

 * mysql+pymysql://root:***@localhost:3306/md_water_services
100 rows affected.


[('HaRu19509224', 'shared_tap', Decimal('3998'), 1),
 ('AkRu05603224', 'shared_tap', Decimal('3998'), 1),
 ('AkRu04862224', 'shared_tap', Decimal('3996'), 3),
 ('AmAs10911224', 'shared_tap', Decimal('3996'), 3),
 ('KiHa22867224', 'shared_tap', Decimal('3996'), 3),
 ('KiZu31330224', 'shared_tap', Decimal('3994'), 6),
 ('HaRu19839224', 'shared_tap', Decimal('3994'), 6),
 ('KiRu28630224', 'shared_tap', Decimal('3992'), 8),
 ('KiZu31415224', 'shared_tap', Decimal('3992'), 8),
 ('AmPw12653224', 'shared_tap', Decimal('3990'), 10),
 ('SoRu38511224', 'shared_tap', Decimal('3990'), 10),
 ('KiRu26218224', 'shared_tap', Decimal('3990'), 10),
 ('HaRu18763224', 'shared_tap', Decimal('3988'), 13),
 ('SoRu35346224', 'shared_tap', Decimal('3988'), 13),
 ('SoRu36081224', 'shared_tap', Decimal('3988'), 13),
 ('SoRu36687224', 'shared_tap', Decimal('3988'), 13),
 ('KiRu29517224', 'shared_tap', Decimal('3986'), 17),
 ('AmAm09725224', 'shared_tap', Decimal('3986'), 17),
 ('SoRu35398224', 'shared_tap', Decimal('3984'), 19),
 ('AmBe11275224', 'shared_tap', Decimal('3984'), 19),
 ('HaRu18154224', 'shared_tap', Decimal('3984'), 19),
 ('SoRu38869224', 'shared_tap', Decimal('3984'), 19),
 ('AkRu05842224', 'shared_tap', Decimal('3984'), 19),
 ('AkRu06647224', 'shared_tap', Decimal('3982'), 24),
 ('KiMr24719224', 'shared_tap', Decimal('3982'), 24),
 ('SoRu35229224', 'shared_tap', Decimal('3982'), 24),
 ('HaRu20823224', 'shared_tap', Decimal('3982'), 24),
 ('KiRu28844224', 'shared_tap', Decimal('3982'), 24),
 ('SoRu37216224', 'shared_tap', Decimal('3982'), 24),
 ('SoRu38339224', 'shared_tap', Decimal('3980'), 30),
 ('AmRu13722224', 'shared_tap', Decimal('3980'), 30),
 ('SoRu35598224', 'shared_tap', Decimal('3980'), 30),
 ('AkRu06374224', 'shared_tap', Decimal('3980'), 30),
 ('KiRu25344224', 'shared_tap', Decimal('3980'), 30),
 ('KiRu25702224', 'shared_tap', Decimal('3978'), 35),
 ('HaRu18222224', 'shared_tap', Decimal('3978'), 35),
 ('AkRu07796224', 'shared_tap', Decimal('3978'), 35),
 ('AmAb09339224', 'shared_tap', Decimal('3978'), 35),
 ('SoRu37693224', 'shared_tap', Decimal('3978'), 35),
 ('KiZu31086224', 'shared_tap', Decimal('3978'), 35),
 ('SoRu35204224', 'shared_tap', Decimal('3976'), 41),
 ('SoRu39073224', 'shared_tap', Decimal('3976'), 41),
 ('SoRu36782224', 'shared_tap', Decimal('3974'), 43),
 ('AkRu03444224', 'shared_tap', Decimal('3974'), 43),
 ('AkRu04657224', 'shared_tap', Decimal('3972'), 45),
 ('AkRu04357224', 'shared_tap', Decimal('3970'), 46),
 ('AkRu03117224', 'shared_tap', Decimal('3970'), 46),
 ('HaRu18399224', 'shared_tap', Decimal('3970'), 46),
 ('KiRu28759224', 'shared_tap', Decimal('3970'), 46),
 ('KiRu29158224', 'shared_tap', Decimal('3970'), 46),
 ('AmRu15466224', 'shared_tap', Decimal('3970'), 46),
 ('KiRu25414224', 'shared_tap', Decimal('3968'), 52),
 ('KiZu31296224', 'shared_tap', Decimal('3968'), 52),
 ('AkRu05208224', 'shared_tap', Decimal('3962'), 54),
 ('SoRu35132224', 'shared_tap', Decimal('3960'), 55),
 ('AkRu08091224', 'shared_tap', Decimal('3958'), 56),
 ('AkRu06749224', 'shared_tap', Decimal('3958'), 56),
 ('HaRu19088224', 'shared_tap', Decimal('3958'), 56),
 ('AkRu03258224', 'shared_tap', Decimal('3956'), 59),
 ('KiRu30056224', 'shared_tap', Decimal('3956'), 59),
 ('AkRu07466224', 'shared_tap', Decimal('3956'), 59),
 ('KiRu30571224', 'shared_tap', Decimal('3954'), 62),
 ('KiRu27212224', 'shared_tap', Decimal('3954'), 62),
 ('AkRu03733224', 'shared_tap', Decimal('3952'), 64),
 ('SoRu34996224', 'shared_tap', Decimal('3952'), 64),
 ('SoRu35794224', 'shared_tap', Decimal('3952'), 64),
 ('AmRu15209224', 'shared_tap', Decimal('3952'), 64),
 ('AmRu13488224', 'shared_tap', Decimal('3952'), 64),
 ('AmAs10579224', 'shared_tap', Decimal('3952'), 64),
 ('AmAs10546224', 'shared_tap', Decimal('3950'), 70),
 ('AmAs10638224', 'shared_tap', Decimal('3950'), 70),
 ('KiRu29396224', 'shared_tap', Decimal('3948'), 72),
 ('SoRu34652224', 'shared_tap', Decimal('3948'), 72),
 ('KiRu27464224', 'shared_tap', Decimal('3946'), 74),
 ('AkRu074592

#### Analysing queues

Ok, these are some of the things I think are worth looking at:
1. How long did the survey take?
2. What is the average total queue time for water?
3. What is the average queue time on different days?
4. How can we communicate this information efficiently?

In [3]:
%%sql
SELECT *
FROM 
    visits
LIMIT 10;    

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46),
 (2, 'HaRu19752', 'HaRu19752224', datetime.datetime(2021, 1, 1, 9, 36), 1, 62, 40),
 (3, 'AkLu01628', 'AkLu01628224', datetime.datetime(2021, 1, 1, 9, 53), 1, 0, 1),
 (4, 'AkRu03357', 'AkRu03357224', datetime.datetime(2021, 1, 1, 10, 11), 1, 28, 14),
 (5, 'KiRu29315', 'KiRu29315224', datetime.datetime(2021, 1, 1, 10, 17), 1, 9, 40),
 (6, 'AkRu05234', 'AkRu05234224', datetime.datetime(2021, 1, 1, 10, 18), 1, 0, 30),
 (7, 'KiRu28520', 'KiRu28520224', datetime.datetime(2021, 1, 1, 10, 28), 1, 0, 34),
 (8, 'HaZa21742', 'HaZa21742224', datetime.datetime(2021, 1, 1, 10, 37), 1, 0, 6),
 (9, 'AmDa12214', 'AmDa12214224', datetime.datetime(2021, 1, 1, 10, 58), 1, 0, 36)]

##### Question 1:
To calculate how long the survey took, we need to get the first and last dates (which functions can find the largest/smallest value), and subtract
them. Remember with DateTime data, we can't just subtract the values. We have to use a function to get the difference in days.

In [9]:
%%sql
SELECT
    MAX(time_of_record) AS end_date,
    MIN(time_of_record) AS start_date,
    TIMESTAMPDIFF(day, MIN(time_of_record), MAX(time_of_record)) AS 'Days elapsed'
FROM
    visits      

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(datetime.datetime(2023, 7, 14, 13, 53), datetime.datetime(2021, 1, 1, 9, 10), 924)]

##### Question 2:
Let's see how long people have to queue on average in Maji Ndogo. Keep in mind that many sources like taps_in_home have no queues. These
are just recorded as 0 in the time_in_queue column, so when we calculate averages, we need to exclude those rows. Try using NULLIF() do to
this.

In [11]:
%%sql
SELECT
    AVG(NULLIF(time_in_queue,0)) AS 'Average queue time'
FROM
    visits;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


[(Decimal('123.2574'),)]

So on average, people take **two hours** to fetch water if they don't have a tap in their homes.

##### Question 3:
So let's look at the queue times aggregated across the different days of the week


In [22]:
%%sql
SELECT
    DAYNAME(time_of_record) AS 'Day of the week',
    ROUND(AVG(NULLIF(time_in_queue,0)))AS 'Average queue time'
FROM
    visits
GROUP BY
    DAYNAME(time_of_record);




 * mysql+pymysql://root:***@localhost:3306/md_water_services
7 rows affected.


[('Friday', Decimal('120')),
 ('Saturday', Decimal('246')),
 ('Sunday', Decimal('82')),
 ('Monday', Decimal('137')),
 ('Tuesday', Decimal('108')),
 ('Wednesday', Decimal('97')),
 ('Thursday', Decimal('105'))]

Wow, ok Saturdays have much longer queue times compared to the other days!


##### Question 4:
We can also look at what time during the day people collect water. Try to order the results in a meaningful way

In [32]:
%%sql
SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS 'Hour of the day',
    ROUND(AVG(NULLIF(time_in_queue,0))) AS 'Average queue time'
FROM
    visits
GROUP BY
    TIME_FORMAT(TIME(time_of_record), '%H:00')
ORDER BY 
    TIME_FORMAT(TIME(time_of_record), '%H:00');




 * mysql+pymysql://root:***@localhost:3306/md_water_services
14 rows affected.


[('06:00', Decimal('149')),
 ('07:00', Decimal('149')),
 ('08:00', Decimal('149')),
 ('09:00', Decimal('118')),
 ('10:00', Decimal('114')),
 ('11:00', Decimal('111')),
 ('12:00', Decimal('112')),
 ('13:00', Decimal('115')),
 ('14:00', Decimal('114')),
 ('15:00', Decimal('114')),
 ('16:00', Decimal('114')),
 ('17:00', Decimal('149')),
 ('18:00', Decimal('147')),
 ('19:00', Decimal('168'))]

In [38]:
%%sql
SELECT
    TIME_FORMAT(TIME(time_of_record), '%H:00') AS hour_of_day,
    -- Sunday
    ROUND(AVG(
        CASE
            WHEN DAYNAME(time_of_record) = 'Sunday' THEN time_in_queue
            ELSE NULL
        END
            ), 0) AS Sunday,
    -- Monday
    ROUND(AVG(
        CASE
            WHEN DAYNAME(time_of_record) = 'Monday' THEN time_in_queue
            ELSE NULL
        END
            ), 0) AS Monday,
    -- Tuesday
    ROUND(AVG(
        CASE
            WHEN DAYNAME(time_of_record) = 'Tuesday' THEN time_in_queue
            ELSE NULL
        END
            ), 0) AS Tuesday,
    
    -- Wednesday
    ROUND(AVG(
        CASE
            WHEN DAYNAME(time_of_record) = 'Wednesday' THEN time_in_queue
            ELSE NULL
        END
            ), 0) AS Wednesday,
    -- Thursday
    ROUND(AVG(
        CASE
            WHEN DAYNAME(time_of_record) = 'Thursday' THEN time_in_queue
            ELSE NULL
        END
            ), 0) AS Thursday,
    -- Friday
    ROUND(AVG(
        CASE
            WHEN DAYNAME(time_of_record) = 'Friday' THEN time_in_queue
            ELSE NULL
        END
            ), 0) AS Friday,
    -- Saturday
    ROUND(AVG(
        CASE
            WHEN DAYNAME(time_of_record) = 'Saturday' THEN time_in_queue
            ELSE NULL
        END
            ), 0) AS Saturday,
    -- Sunday
    ROUND(AVG(
        CASE
            WHEN DAYNAME(time_of_record) = 'Sunday' THEN time_in_queue
            ELSE NULL
        END
            ), 0) AS Sunday
FROM
visits
WHERE
time_in_queue != 0 -- this excludes other sources with 0 queue times
GROUP BY
hour_of_day
ORDER BY
hour_of_day;                  

 * mysql+pymysql://root:***@localhost:3306/md_water_services
14 rows affected.


[('06:00', Decimal('79'), Decimal('190'), Decimal('134'), Decimal('112'), Decimal('134'), Decimal('153'), Decimal('247'), Decimal('79')),
 ('07:00', Decimal('82'), Decimal('186'), Decimal('128'), Decimal('111'), Decimal('139'), Decimal('156'), Decimal('247'), Decimal('82')),
 ('08:00', Decimal('86'), Decimal('183'), Decimal('130'), Decimal('119'), Decimal('129'), Decimal('153'), Decimal('247'), Decimal('86')),
 ('09:00', Decimal('84'), Decimal('127'), Decimal('105'), Decimal('94'), Decimal('99'), Decimal('107'), Decimal('252'), Decimal('84')),
 ('10:00', Decimal('83'), Decimal('119'), Decimal('99'), Decimal('89'), Decimal('95'), Decimal('112'), Decimal('259'), Decimal('83')),
 ('11:00', Decimal('78'), Decimal('115'), Decimal('102'), Decimal('86'), Decimal('99'), Decimal('104'), Decimal('236'), Decimal('78')),
 ('12:00', Decimal('78'), Decimal('115'), Decimal('97'), Decimal('88'), Decimal('96'), Decimal('109'), Decimal('239'), Decimal('78')),
 ('13:00', Decimal('81'), Decimal('122'), Decimal('97'), Decimal('98'), Decimal('101'), Decimal('115'), Decimal('242'), Decimal('81')),
 ('14:00', Decimal('83'), Decimal('127'), Decimal('104'), Decimal('92'), Decimal('96'), Decimal('110'), Decimal('244'), Decimal('83')),
 ('15:00', Decimal('83'), Decimal('126'), Decimal('104'), Decimal('88'), Decimal('92'), Decimal('110'), Decimal('248'), Decimal('83')),
 ('16:00', Decimal('83'), Decimal('127'), Decimal('99'), Decimal('90'), Decimal('99'), Decimal('109'), Decimal('251'), Decimal('83')),
 ('17:00', Decimal('79'), Decimal('181'), Decimal('135'), Decimal('121'), Decimal('129'), Decimal('151'), Decimal('251'), Decimal('79')),
 ('18:00', Decimal('80'), Decimal('174'), Decimal('122'), Decimal('113'), Decimal('132'), Decimal('158'), Decimal('240'), Decimal('80')),
 ('19:00', Decimal('127'), Decimal('159'), Decimal('145'), Decimal('176'), Decimal('137'), Decimal('103'), Decimal('282'), Decimal('127'))]

I have spoted these patterns:
1. Queues are very long on a Monday morning and Monday evening as people rush to get water.
2. Wednesday has the lowest queue times, but long queues on Wednesday evening.
3. People have to queue pretty much twice as long on Saturdays compared to the weekdays. It looks like people spend their Saturdays queueing
for water, perhaps for the week's supply?
4. The shortest queues are on Sundays, and this is a cultural thing. The people of Maji Ndogo prioritise family and religion, so Sundays are spent
with family and friends.

In [41]:
import sqlite3
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# Corrected file path
file_path = r"D:\Data Science-Explore AI course\SQL\12.Integrated Project-Clustering Data To Unveil Maji Ndogo's Water Crisis\md_water_services.sql"

# Read the SQL text file
with open(file_path, 'r') as file:
    sql_script = file.read()

# Connect to an in-memory SQLite database
conn = sqlite3.connect(':memory:')
cursor = conn.cursor()

# Execute the SQL script
cursor.executescript(sql_script)

# Query to get the data
query = """
SELECT 
    time_of_record,
    time_in_queue
FROM 
    visits
WHERE 
    time_in_queue != 0
"""
df = pd.read_sql_query(query, conn)

# Convert time_of_record to datetime
df['time_of_record'] = pd.to_datetime(df['time_of_record'])

# Extract day of the week and hour
df['day_of_week'] = df['time_of_record'].dt.day_name()
df['hour_of_day'] = df['time_of_record'].dt.hour

# Calculate average queue time by day of the week and hour
avg_queue_time = df.groupby(['day_of_week', 'hour_of_day'])['time_in_queue'].mean().unstack()

# Plotting
plt.figure(figsize=(14, 8))
sns.heatmap(avg_queue_time, cmap='YlOrRd', annot=True, fmt=".1f", linewidths=.5)
plt.title('Queue times for different days and hours')
plt.xlabel('Hour of the Day')
plt.ylabel('Day of the Week')
plt.show()

# Close the connection
conn.close()


OperationalError: near "DATABASE": syntax error